Add AuthorizationManager to Messaging

jzheaux opened the pull request:
Instead of extending `AbstractSecurityWebSocketMessageBrokerConfigurer`, folks can do:

```java
@EnableWebSocketSecurity
@Configuration
public class SecurityConfig {
    @Bean  
    AuthorizationManager<Message<?>> authorizationManager(
            MessageMatcherDelegatingAuthorizationManager.Builder messages) {
        messages
            .simpDestMatchers("/path/**").permitAll()
            .anyMessage().denyAll();
        return messages.build();
    }
}
```

rwinch commented on config/src/main/java/org/springframework/security/config/annotation/web/configuration/EnableWebSecurity.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2002-2020 the original author or authors.
+ * Copyright 2002-2022 the original author or authors.
```
I think this can be reverted

rwinch commented on config/src/main/java/org/springframework/security/config/annotation/web/messaging/AuthorizationManagerMessageMatcherRegistry.java:
```diff
@@ -0,0 +1,354 @@
+		return new Constraint(builders);
+	}
+
+	public AuthorizationManager<Message<?>> build() {
```
I think this is more of a builder than a registry since it exposes a build method. It makes me wonder if we are building an AuthorizationManager if we out to just provide a builder on MessageMatcherDelegatingAuthorizationManager as you mentioned. We'd need to double check for package tangles though.

rwinch commented on config/src/main/java/org/springframework/security/config/annotation/web/socket/AbstractSecurityWebSocketMessageBrokerConfigurer.java:
```diff
@@ -81,9 +81,15 @@
  * @since 4.0
+ * @see WebSocketMessageBrokerSecurityConfiguration
+ * @deprecated Use
+ * {@link org.springframework.security.config.annotation.web.configuration.EnableWebSecurity}
```
Should this be EnableWebSocketSecurity?

rwinch commented on config/src/main/java/org/springframework/security/config/annotation/web/messaging/AuthorizationManagerMessageMatcherRegistry.java:
```diff
@@ -0,0 +1,354 @@
+
+	private PathMatcher pathMatcher;
+
+	public AuthorizationManagerMessageMatcherRegistry(ApplicationContext context) {
```
Does this need to be public or can it be package scope?

jzheaux replied:
Removed class.

rwinch commented on config/src/main/java/org/springframework/security/config/annotation/web/messaging/AuthorizationManagerMessageMatcherRegistry.java:
```diff
@@ -0,0 +1,354 @@
+	private final MessageMatcherDelegatingAuthorizationManager.Builder builder = MessageMatcherDelegatingAuthorizationManager
+			.builder();
+
+	private final ApplicationContext context;
```
Does it make more sense to always inject PathMatcher rather than use `ApplicationContext`. If we need to delay the lookup perhaps a Supplier<PathMatcher>`.

jzheaux replied:
The reason for deferring is to solve a dependency cycle. Calling for the `PathMatcher` instance from the `ApplicationContext` causes the `DelegatingWebSocketMessageBrokerConfiguration` to be constructed before this configurer can be registered to it.

I'll change it to a `Supplier` and we can go from there.

rwinch commented on config/src/main/java/org/springframework/security/config/annotation/web/socket/WebSocketMessageBrokerSecurityConfiguration.java:
```diff
@@ -0,0 +1,171 @@
+ * 	&#064;Bean
+ * 	AuthorizationManager&lt;Message&lt;?&gt;&gt; (AuthorizationManagerMessageMatcherRegistry messages) {
+ * 		messages.simpDestMatchers(&quot;/user/queue/errors&quot;).permitAll()
+ * 				.simpDestMatchers(&quot;/admin/**&quot;).hasRole(&quot;ADMIN&quot;).anyMessage()
```
I wonder if this could be better formatted with anyMessage on a new line

rwinch commented on config/src/main/java/org/springframework/security/config/annotation/web/socket/WebSocketMessageBrokerSecurityConfiguration.java:
```diff
@@ -0,0 +1,171 @@
+ *
+ * <pre>
+ * &#064;Configuration
+ * &#064;EnableWebSecurity
```
Does it need `@EnableWebSocketSecurity`

jzheaux replied:
Removed JavaDoc since package-private class.

rwinch commented on config/src/main/java/org/springframework/security/config/annotation/web/socket/WebSocketMessageBrokerSecurityConfiguration.java:
```diff
@@ -0,0 +1,171 @@
+ * For example:
+ * </p>
+ *
+ * <pre>
```
It might be better to have this documented on `@EnableWebSocketSecurity` since it is public while this class is not.

rwinch commented on messaging/src/main/java/org/springframework/security/messaging/access/intercept/AuthorizationChannelInterceptor.java:
```diff
@@ -0,0 +1,95 @@
+	public Message<?> preSend(Message<?> message, MessageChannel channel) {
+		this.logger.debug(LogMessage.of(() -> "Authorizing message send"));
+		AuthorizationDecision decision = this.preSendAuthorizationManager.check(AUTHENTICATION_SUPPLIER, message);
+		this.eventPublisher.publishAuthorizationEvent(AUTHENTICATION_SUPPLIER, message, decision);
```
I think that eventPublisher can be null. Perhaps it would be good to default eventPublisher to a no-op implementation.

rwinch commented on messaging/src/main/java/org/springframework/security/messaging/util/matcher/MessageMatcherEntry.java:
```diff
@@ -0,0 +1,44 @@
+ * @author Josh Cummings
+ * @since 5.7
+ */
+public class MessageMatcherEntry<T> {
```
What are the benefits of this being a top level class vs inner class of MessageMatcherDelegatingAuthorizationManager

jzheaux replied:
Inlined.

rwinch reviewed (changes requested):
Thanks for the pull request. I have provided feedback inline. Please also review for package tangles.

jzheaux commented:
@rwinch I added XML configuration and responded to your other points.

Note that regarding XML configuration, I originally coded it with an opt-out feature to revert back to using `SecurityMetadataSource`. However, I couldn't find any reason to actually use the feature in an existing application, so I removed it. If you see a reason to add the flag now, this can be added easily as I have the code preserved in an earlier commit in the PR.

jzheaux commented:
Update: Because of #9667, I think that it's important to still allow applications to opt-in to using `AuthorizationManager`. The reason is that #9667 cannot call the `Authentication`-based `protected` methods in `AbstractSecurityExpressionHandler`. If applications have overridden those, then their custom code would stop being called when they upgrade.

So, the PR now introduces `use-authorization-manager`. When the value is `true` or when the application uses the `authorization-manager-ref` attribute, Spring Security will use the `AuthorizationManager` API. In Spring Security 6.0, this attribute can be deprecated and defaulted to `true`.

rwinch commented on core/src/main/java/org/springframework/security/access/expression/SecurityExpressionHandler.java:
```diff
@@ -42,7 +43,10 @@
 	 * invocation type.
 	 */
-	EvaluationContext createEvaluationContext(Authentication authentication, T invocation);
+	default EvaluationContext createEvaluationContext(Authentication authentication, T invocation) {
```
Is there a reason to default this method in the same commit?

jzheaux replied:
Shoot, good catch! It's left over from an experiment. I'll revert.

rwinch reviewed (changes requested):
Thanks. I left a comment inline.
